src/transformStateWithClones.js

```diff
@@ -7,7 +7,32 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let currentState = { ...state };
+  const stateHistory = [];
+
+  for (let i = 0; i < actions.length; i++) {
+    const action = actions[i];
+    let nextState = { ...currentState };
+
+    switch (action.type) {
+      case 'clear':
+        nextState = {};
+        break;
+      case 'addProperties':
+        Object.assign(nextState, action.extraData);
+        break;
+      case 'removeProperties':
+        action.keysToRemove.forEach((key) => delete nextState[key]);
+        break;
+      default:
+        break;
+    }
```

Comment:
Consider using a `default` case in your switch statement that throws an error or logs a message. This will help with error handling and ensure that any unexpected `action.type` values are caught.